README.txt

CONTENTS OF THIS FILE
---------------------

 * Introduction
 * Available scripts
 * Installing your own Readmilling


INTRODUCTION
------------

Current Maintainer: Morbus Iff <morbus@disobey.com>

Readmilling is a collection of potentially useful open source scripts using
the API provided by Readmill, a social ebook reader available for iPhone and
iPad. These scripts are not affiliated with or created by the fine chaps and
chapettes at Readmill.com.

  Readmill:                 http://readmill.com/
  Readmilling Online:       http://www.disobey.com/d/code/readmilling/
  Readmilling source code:  https://github.com/morbus/readmilling/

Found bugs or have a request? Like 'em and want to give thanks?

  Email:                    morbus@disobey.com
  Twitter:                  https://twitter.com/morbusiff
  Amazon wishlist:          http://www.amazon.com/gp/registry/25USVJDH68554


AVAILABLE SCRIPTS
-----------------

 * book-comments.php tries to merge all comments left on similar highlights
   into a single entry. Different versions of the same book can have highlights
   in different locations, so the script can't guarantee the merged entries
   are in the same order they appear in the text.


INSTALLING YOUR OWN READMILLING
-------------------------------

There are two ways to run Readmilling:

 * Through someone else's installation
 * Through your own installation

The quickest approach is through the official web installation at:

  http://www.disobey.com/d/code/readmilling/

If you'd like to customize the display of your data with a new theme, or
else tweak Readmilling's code, you'll need your own installation with:

 * PHP 5.2+ with HTTP_Request2 from PEAR.
 * A Readmill API key from http://developers.readmill.com/
 * A web server with support for PHP scripts.

Grab the source from https://github.com/morbus/readmilling, then:

 * Specify your API key into the settings.php (or settings.local.php).
 * Create a directory called cache/ in the root of the Readmilling source.
 * Give your webserver permission to write to this cache/ directory.

You should now be able to access index.html and the scripts from scripts/.

Don't hesitate to email morbus@disobey.com with further questions.



CHANGELOG.txt

BUGS AND TODOS
--------------
general:
  * @todo this day in readmill script
  * @todo advanced search engine
  * @bug center small book images: http://www.disobey.com/detergent/code/readmilling/scripts/book-comments.php?title=accelerando&author=charles+stross
  * @todo consider switching the default books to literature instead of technocentric.
  * @todo we probably need to htmlentities everything we’re spitting out to be safe.
  * @todo And see what you can do about moving them to htmlspecialchars instead.
  * @todo Fix display of all scripts when bad API key or no data returned.
  * @todo add Morbus Iff block to add pages.
  * @todo add full name matching in user/search (for david kjelkerud).
  * @todo replace “Be aware” error messages with hidable progress bar?

user-export.php:
  * @todo add to changelog and readme
  * @todo add to front page
  * reading periods are not supported yet.
  * closing comments are not supported yet.
  * ideally should use oAuth for private data export.
  * wd be great to have OPML (Highlights—>Comments) so I can pull into iThoughts
    https://twitter.com/adoran2/status/412961190919553024

book-comments.php:
  * @todo add closing comments
  * @todo add a no_merging=1 flag for debugging/example purposes.
  * @todo replace our custom comment permalink with highlight permalink?
  * @todo No data message?

book-updates.php:
  * @todo add to changelog and readme
  * @todo add to front page
  * @todo add closing comments
  * @todo No data message?

user-updates.php:
  * @todo add to changelog and readme
  * @todo add to front page
  * @todo write up sidebar intro
  * @todo add RSS export and icon/sniff thing.
  * @todo fill in your @todos. heh.
  * @todo No data message?
  * @todo note new api hooks for user_ lookups.


VERSION 0.2, 2014-XX-XX
-----------------------
  * Removed a ton of duplicate code from common.inc.
  * All settings have been changed to hyphens, not underscores.
  * Switched over to Readmill's "pagination" API support.
  * Fetch all highlights per reading, not just the first 100.
  * Scripts can now (dangerously) ignore caching for certain requests.
  * All caches now default to expiring after one day. We'll try it out.


VERSION 0.1, 2013-02-22
-----------------------
  * Initial public release.



scripts/user-export.php
<?php

/**
 * @file
 * Export a Readmill user's data.
 */

include_once "../includes/common.inc";
disable_output_buffering();

$username = isset($_REQUEST['username']) ? $_REQUEST['username'] : '';

// Retrieve, store, package.
function start_export($username) {
  $data = new stdClass;

  // @todo Check for cache time and prevent regeneration.
  // Should we need to, given that the other files are cached?
  // If anything, we'd be saving zip file generation only.

  if (!$data->user = readmill_user_search($username)) {
    // @todo turn this into existing error template.
    print 'The requested user could not be found.';
    return;
  }

  // Store here for packaging.
  $export_dir = '../exports/';
  $export_user_dir = $export_dir . $username . '/';
  @mkdir($export_user_dir, 0755, TRUE);

  print "<ul><li>Requesting readings, highlights, and comments: ";
  $data->readings = readmill_user_readings($data->user->id);
  foreach ($data->readings as $rid => $reading) {
    // Remove duplicated user information. Readmill's API will copy this
    // into individual readings, highlights, and comments but we'll remove
    // it for the sake of a cleaner and smaller export. In the relatively
    // small test case of 'morbus', this reduced file size by roughly 35%.
    unset($reading->user);

    // @todo Doesn't support reading periods yet.

    if ($reading->highlights_count >= 1) {
      $data->readings->$rid->highlights = readmill_reading_highlights($reading->id);
      foreach ($data->readings->$rid->highlights as $hid => $highlight) {
        unset($highlight->user);

        if ($highlight->comments_count >= 1) {
          $data->readings->$rid->highlights->$hid->comments = readmill_highlight_comments($highlight->id);
          foreach ($data->readings->$rid->highlights->$hid->comments as $cid => $comment) {
            print heartbeat();
            unset($comment->user);
          }
        }
      }
    }
  }

  // For ease of parsing by developers, a giant dump of everything combined.
  file_put_contents($export_user_dir . 'all-user-data.json', json_encode($data), LOCK_EX);

  // Create some friendlier exports.
  print " Generating exports in Markdown: ";
  foreach ($data->readings as $reading) {
    print heartbeat();
    $safe_title = preg_replace('/[^a-zA-Z0-9 ]+/', '', $reading->book->title);
    $export_book_dir = $export_user_dir . 'library/' . $safe_title . '/';
    @mkdir($export_book_dir, 0755, TRUE);

    $markdown = "**" . $reading->book->title . '** by *' . $reading->book->author . "*\n\n";

    if (!empty($reading->highlights)) {
      foreach ($reading->highlights as $highlight) {
        $markdown .= $highlight->content . "\n\n---\n\n";
      }
    }

    file_put_contents($export_book_dir . $safe_title . '.md', $markdown, LOCK_EX);
  }
  print ' done.</li></ul>';

  // Zip and link. Yeah, there's a PHP ZipArchive extension, but it'd require
  // about twenty lines of code to handle recursion and deletion and meh.
  if (exec('cd ' . escapeshellarg($export_dir) . ' && zip -mrT -FS ' . escapeshellarg($username . '.zip') . ' ' . escapeshellarg($username))) {
    print '<p>Your Readmill data has been exported and is <a href="../exports/' . $username . '.zip">now available for download</a>.</p>';
  }
  else {
    print '<p>There was a failure collecting your data. E-mail <a href="mailto:morbus@disobey.com">morbus@disobey.com</a> for further support.</p>';
  }
}
?>
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <title>User data export - Readmilling</title>
  <link rel="stylesheet" href="../misc/default.css" />
</head>
<body>
  <header class="page-toolbar toolbar">
    <div class="container">
      <div class="readmilling-explanation">
        <a href="http://www.disobey.com/d/code/readmilling/">Readmilling</a> is
        a <a href="https://github.com/morbus/readmilling">collection of scripts</a> for use with
        <a href="http://readmill.com/">Readmill's</a> API. Created by <a href="http://www.disobey.com/">Morbus Iff</a>.
        <span class="backpedaling">These scripts are not affiliated with or created by the fine chaps and chapettes at Readmill.com.</span>
      </div>
    </div>
  </header>

  <div class="page-content container">
    <?php if (!empty($username)) { ?>
    <section class="progress-report">
      <h1>Progress report</h1>
      <?php start_export($username); ?>
    </section>
    <?php } ?>

    <section class="layout-primary">
      <div class="layout-primary-column">
        <header>
          <h1>
            <span class="header-group-primary">User data export</span>
            <span class="header-group-secondary">You put it in, now get it out.</span>
          </h1>
        </header>
        <section id="user-export-form">
          <form action="user-export.php" accept-charset="UTF-8" method="get">
            <label for="form-title">Username</label><input id="form-title" name="username" type="text" placeholder="<?php print htmlentities($username, ENT_COMPAT, "UTF-8"); ?>" required />
            <button>Export user data</button> <!-- NP: 'Let's Go Away (Daytona USA)' from Various Artists's album 'Segacon: The Best Of Sega Game Music Vol. 1' -->
          </form>
          <span class="warning"><strong>Be aware:</strong> If we've not seen this user before,
          or its data has expired, it might take a few minutes before you'll get results.</span>
        </section>

        <section id="user-export-explanation">
          <h1>What you'll get</h1>

        </section>
      </div>
    </section>

    <aside class="layout-secondary">
      <div class="layout-secondary-column">
        <section class="secondary-section">
          <h1>About this script</h1>
          <p>@todo</p>
        </section>
        <section class="secondary-section">
          <h1>Morbus Iff</h1>
          <p>Found bugs or have a request? Like it and want to give thanks?</p>
          <ul>
            <li><a href="https://twitter.com/morbusiff">@morbusiff</a></li>
            <li>E-mail and PayPal: <a href="mailto:morbus@disobey.com">morbus@disobey.com</a></li>
            <li><a href="http://www.amazon.com/gp/registry/25USVJDH68554">Amazon wishlist</a></li>
            <li><a href="http://www.disobey.com/">Disobey.com</a></li>
          </ul>
        </section>
      </div>
    </aside>
  </div>

  <footer>
    <div class="container">
      <a href="http://www.disobey.com/d/code/readmilling/">Readmilling</a> is free software released
      under the <a href="http://www.gnu.org/licenses/quick-guide-gplv3.html">GPL v3</a> or later.
      <a href="https://github.com/morbus/readmilling">Browse the source code at Github.</a>
    </div>
  </footer>
</body>
</html>


scripts/user-updates.php
<?php

/**
 * @file
 * View a user's most recent highlights and comments.
 */

include_once("../includes/common.inc");

// Provide defaults.
$default_username = 'morbus';
$username  = !empty($_REQUEST['username'])
  ? $_REQUEST['username'] : $default_username;

// Override the defaults if we've been passed user data.
$errors = array(); // YOU"VE GONE AND DONE SOMETHING FOOLISH LUCY.

// Grab *something*.
$data = new stdClass;
$data->user = readmill_user_search($username);
if (!$data->user) { // Fallback to a known example.
  $errors[] = 'The requested user could not be found.';
  $data->user = readmill_user_search($default_username);
}

if (count($errors)) { // Transmogrificate any errors into a renderable string.
  $errors = '<ul class="errors"><li>' . implode('</li><li>', $errors) . '</li></ul>';
}

// Pre-build the URLs of the HTML and RSS versions of this view.
$query_string = '?username=' . urlencode($data->user->username);
$url_for_rss  = 'http://' . $_SERVER['HTTP_HOST'] . $_SERVER['SCRIPT_NAME'] . $query_string . '&rss=1';
$url_for_html = 'http://' . $_SERVER['HTTP_HOST'] . $_SERVER['SCRIPT_NAME'] . $query_string; // BOO!

// Fetch all the readings for this user. We're interested in readings
// with highlights (and/or comments), but we also want closing_remarks.
if ($data->user->id) { // Occasional server errors. Sanity.
  $data->readings = readmill_user_readings($data->user->id);
}

// For every reading, fetch all the highlights and comments.
foreach ($data->readings as $reading) {
  if ($reading->highlights_count >= 1) {
    $highlights = readmill_reading_highlights($reading->id);

    foreach ($highlights as $highlight) {
      if ($highlight->comments_count >= 1) {
        $comments = readmill_highlight_comments($highlight->id);

        foreach ($comments as $comment) {
          $data->updates[strtotime($comment->posted_at)][] = array(
            'user'            => $comment->user,
            'highlight'       => $highlight,
            'comment'         => $comment,
            'action'          => 'left a comment',
            'content'         => $comment->content,
            'permalink_url'   => $highlight->permalink_url,
            'permalink_text'  => 'Reply to comment',
          );
        }
      }
      else { // No comment was left, just a happy highlight.
        $data->updates[strtotime($highlight->highlighted_at)][] = array(
          'user'            => $highlight->user,
          'highlight'       => $highlight,
          'action'          => 'shared a highlight',
          'content'         => 'shared a highlight.',
          'permalink_url'   => $highlight->permalink_url,
          'permalink_text'  => 'View highlight',
        );
      }
    }
  }
}

krsort($data->updates); // Spit out only the latest 50.
$data->updates = array_slice($data->updates, 0, 50, TRUE);

// Want RSS? HAVE MANUALLY WRTITEN RSS OOOOH YEAAAAAHH.
if (isset($_REQUEST['rss']) && $_REQUEST['rss'] == 1) {
  header('Content-Type: application/rss+xml');
  print '<?xml version="1.0" encoding="utf-8"?>' . "\n";
  print '<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom"><channel>';
  print   '<title>Latest updates for ' . htmlspecialchars($data->user->username, ENT_COMPAT, "UTF-8") . '- Readmilling</title>';
  print   '<description>User updates shows the 50 most recent highlights and comments made by a user in HTML or RSS.</description>';
  print   '<atom:link href="' . htmlspecialchars($url_for_rss, ENT_COMPAT, "UTF-8") .'" rel="self" type="application/rss+xml" />';
  print   '<link>http://' . htmlspecialchars($url_for_html, ENT_COMPAT, "UTF-8") . '</link>';
  print   '<language>en</language>';
  foreach ($data->updates as $timestamp => $updates) {
    foreach ($updates as $update) {
      print '<item>'; // SORT BY LENGTH, G-STRINGS HO!
      print   '<link>' . $update['permalink_url'] . '</link>';
      print   '<pubDate>' . date(DATE_RFC2822, $timestamp) . '</pubDate>';
      print   '<title>' . htmlspecialchars($update['user']->fullname, ENT_COMPAT, "UTF-8") . ' ' . $update['action'] . '</title>';
      print   '<description>';
      print      htmlspecialchars('<blockquote>' . $update['highlight']->content . '</blockquote>', ENT_COMPAT, "UTF-8");
      print      isset($update['comment']) ? htmlspecialchars($update['content'], ENT_COMPAT, "UTF-8") : '';
      print   '</description>';
      print   '<guid isPermaLink="false">' . date(DATE_RFC2822, $timestamp) . ' @ ' . $update['permalink_url'] . '</guid>';
      print '</item>';
    }
  }
  print '</channel></rss>';
  exit; // OH NOESA WINER WINS.
}
?>
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <title>Latest updates for <?php print $data->user->username; ?> - Readmilling</title>
  <link rel="alternate" type="application/rss+xml" title="RSS" href="<?php print htmlentities($url_for_rss, ENT_COMPAT, "UTF-8"); ?>" />
  <link rel="stylesheet" href="../misc/default.css" />
</head>
<body>
  <header class="page-toolbar toolbar">
    <div class="container">
      <div class="readmilling-explanation">
        <a href="http://www.disobey.com/d/code/readmilling/">Readmilling</a> is
        a <a href="https://github.com/morbus/readmilling">collection of scripts</a> for use with
        <a href="http://readmill.com/">Readmill's</a> API. Created by <a href="http://www.disobey.com/">Morbus Iff</a>.
        <span class="backpedaling">These scripts are not affiliated with or created by the fine chaps and chapettes at Readmill.com.</span>
      </div>
    </div>
  </header>

  <div class="page-content container">
    <?php print $errors ? $errors : NULL; ?>
    <section class="layout-primary">
      <div class="layout-primary-column">
        <header>
          <h1>
            <span class="header-group-primary"><a href="<?php print $data->book->permalink_url; ?>"><?php print $data->book->title; ?></a></span>
            <span class="header-group-secondary"><?php print $data->book->author; ?></span>
          </h1>
        </header>
        <section id="highlights">
          <h1>Latest updates</h1>
          <?php
            foreach ($data->updates as $timestamp => $updates) {
              foreach ($updates as $update) {
                print '<article class="highlight">';
                print   '<blockquote>' . htmlentities($update['highlight']->content, ENT_COMPAT, "UTF-8") . '</blockquote>';
                print   '<section class="comments">'; // Be sure to enforce UTF-8 on all htmlentities() due to PHP 5.3 defaults.
                print     '<article class="comment">';
                print       '<a href="' . $update['user']->permalink_url . '" class="image">';
                print         '<img src="' . $update['user']->avatar_url . '" alt="' . $update['user']->fullname . '" />';
                print       '</a>';
                print       '<div class="content">';
                print         '<a href="' . $update['user']->permalink_url . '" class="fullname">' . $update['user']->fullname . '</a> ';
                print         '<p>' . htmlentities($update['content'], ENT_COMPAT, "UTF-8") . '</p>';
                print         '<aside class="metadata">';
                print           '<time class="timestamp" datetime="' . date(DATE_ISO8601, $timestamp) . '">' . date('D, d M Y h:i:s a', $timestamp) . '</time> &middot; ';
                print           '<a href="' . $update['permalink_url'] . '">' . $update['permalink_text'] . '</a>';
                print         '</aside>';
                print       '</div>';
                print     '</article>';
                print   '</section>';
                print '</article>';
              }
            }
          ?>
        </section>
      </div>
    </section>

    <aside class="layout-secondary">
      <div class="layout-secondary-column">
        <section class="book-cover">
          <a href="<?php print $data->book->permalink_url; ?>"><img class="book-cover" alt="Book cover for <?php print $data->book->title; ?>" src="<?php print str_replace('medium', 'original', $data->book->cover_url); ?>" /></a>
        </section>
        <section class="secondary-section">
          <form action="user-updates.php" accept-charset="UTF-8" method="get">
            <label for="form-title">Username</label><input id="form-title" name="username" type="text" placeholder="<?php print htmlentities($username, ENT_COMPAT, "UTF-8"); ?>" required />
            <button>Load user</button>
          </form>
          <span class="warning"><strong>Be aware:</strong> If we've not seen this user before,
          or its data has expired, it might take 30 seconds or more before you'll get results.</span>
        </section>
        <section class="secondary-section">
          <h1>About this script</h1>
          <p><b>User updates</b> shows the 50 most recent highlights and comments made by a user in HTML
          or <a href="<?php print htmlentities($url_for_rss, ENT_COMPAT, "UTF-8"); ?>">RSS</a>. For a
          book-centric approach, see <a href="book-updates.php">Book updates</a>.</p>
        </section>
      </div>
    </aside>
  </div>

  <footer>
    <div class="container">
      <a href="http://www.disobey.com/d/code/readmilling/">Readmilling</a> is free software released
      under the <a href="http://www.gnu.org/licenses/quick-guide-gplv3.html">GPL v3</a> or later.
      <a href="https://github.com/morbus/readmilling">Browse the source code at Github.</a>
    </div>
  </footer>
</body>
</html>


scripts/book-updates.php
<?php

/**
 * @file
 * View a book's most recent highlights and comments.
 */

include_once("../includes/common.inc");

// Provide representational defaults.
$default_match_title  = 'Service Design: From Insight to Implementation';
$default_match_author = 'Andrew Polaine';
$match_title  = $default_match_title;
$match_author = $default_match_author;

// Override the defaults if we've been passed user data.
$errors = array(); // ALL THE INPUTS DAMN IT ALL THE INPUTS.
if (!empty($_REQUEST['title']) && empty($_REQUEST['author'])
   || !empty($_REQUEST['author']) && empty($_REQUEST['title'])) {
  $errors[] = 'Book title and author are both required.';
}
elseif (!empty($_REQUEST['title']) && !empty($_REQUEST['author'])) {
  $match_title  = $_REQUEST['title'];
  $match_author = $_REQUEST['author'];
}

// Grab *something*.
$data = new stdClass;
$data->book = readmill_book_match(array(
  'title'   => $match_title,
  'author'  => $match_author,
));
if (!$data->book) { // Fallback to a known and working example.
  $errors[] = 'The requested book could not be found.';
  $data->book = readmill_book_match(array(
    'title'   => $default_match_title,
    'author'  => $default_match_author,
  ));
}

if (count($errors)) { // Transmogrificate any errors into a renderable string.
  $errors = '<ul class="errors"><li>' . implode('</li><li>', $errors) . '</li></ul>';
}

// Pre-build the URLs of the HTML and RSS versions of this view.
$query_string = '?title=' . urlencode($data->book->title) . '&author=' . urlencode($data->book->author);
$url_for_rss  = 'http://' . $_SERVER['HTTP_HOST'] . $_SERVER['SCRIPT_NAME'] . $query_string . '&rss=1';
$url_for_html = 'http://' . $_SERVER['HTTP_HOST'] . $_SERVER['SCRIPT_NAME'] . $query_string; // BOO!

// Fetch all the readings for this book. We're interested in readings
// with highlights (and/or comments), but we also want closing_remarks.
$data->readings = readmill_book_readings($data->book->id);

// For every reading, fetch all the highlights and comments.
foreach ($data->readings as $reading) {
  if ($reading->highlights_count >= 1) {
    $highlights = readmill_reading_highlights($reading->id);

    foreach ($highlights as $highlight) {
      if ($highlight->comments_count >= 1) {
        $comments = readmill_highlight_comments($highlight->id);

        foreach ($comments as $comment) {
          $data->updates[strtotime($comment->posted_at)][] = array(
            'user'            => $comment->user,
            'highlight'       => $highlight,
            'comment'         => $comment,
            'action'          => 'left a comment',
            'content'         => $comment->content,
            'permalink_url'   => $highlight->permalink_url,
            'permalink_text'  => 'Reply to comment',
          );
        }
      }
      else { // No comment was left, just a happy highlight.
        $data->updates[strtotime($highlight->highlighted_at)][] = array(
          'user'            => $highlight->user,
          'highlight'       => $highlight,
          'action'          => 'shared a highlight',
          'content'         => 'shared a highlight.',
          'permalink_url'   => $highlight->permalink_url,
          'permalink_text'  => 'View highlight',
        );
      }
    }
  }
}

krsort($data->updates); // Spit out only the latest 50.
$data->updates = array_slice($data->updates, 0, 50, TRUE);

// Want RSS? HAVE MANUALLY WRTITEN RSS OOOOH YEAAAAAHH.
if (isset($_REQUEST['rss']) && $_REQUEST['rss'] == 1) {
  header('Content-Type: application/rss+xml');
  print '<?xml version="1.0" encoding="utf-8"?>' . "\n";
  print '<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom"><channel>';
  print   '<title>Book updates for ' . htmlspecialchars($data->book->title, ENT_COMPAT, "UTF-8") . '- Readmilling</title>';
  print   '<description>Book updates shows the 50 most recent highlights and comments made to a book in HTML or RSS.</description>';
  print   '<atom:link href="' . htmlspecialchars($url_for_rss, ENT_COMPAT, "UTF-8") .'" rel="self" type="application/rss+xml" />';
  print   '<link>http://' . htmlspecialchars($url_for_html, ENT_COMPAT, "UTF-8") . '</link>';
  print   '<language>en</language>';
  foreach ($data->updates as $timestamp => $updates) {
    foreach ($updates as $update) {
      print '<item>'; // SORT BY LENGTH, G-STRINGS HO!
      print   '<link>' . $update['permalink_url'] . '</link>';
      print   '<pubDate>' . date(DATE_RFC2822, $timestamp) . '</pubDate>';
      print   '<title>' . htmlspecialchars($update['user']->fullname, ENT_COMPAT, "UTF-8") . ' ' . $update['action'] . '</title>';
      print   '<description>';
      print      htmlspecialchars('<blockquote>' . $update['highlight']->content . '</blockquote>', ENT_COMPAT, "UTF-8");
      print      isset($update['comment']) ? htmlspecialchars($update['content'], ENT_COMPAT, "UTF-8") : '';
      print   '</description>';
      print   '<guid isPermaLink="false">' . date(DATE_RFC2822, $timestamp) . ' @ ' . $update['permalink_url'] . '</guid>';
      print '</item>';
    }
  }
  print '</channel></rss>';
  exit; // OH NOESA WINER WINS.
}
?>
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <title>Book updates for <?php print $data->book->title; ?> - Readmilling</title>
  <link rel="alternate" type="application/rss+xml" title="RSS" href="<?php print htmlentities($url_for_rss, ENT_COMPAT, "UTF-8"); ?>" />
  <link rel="stylesheet" href="../misc/default.css" />
</head>
<body>
  <header class="page-toolbar toolbar">
    <div class="container">
      <div class="readmilling-explanation">
        <a href="http://www.disobey.com/d/code/readmilling/">Readmilling</a> is
        a <a href="https://github.com/morbus/readmilling">collection of scripts</a> for use with
        <a href="http://readmill.com/">Readmill's</a> API. Created by <a href="http://www.disobey.com/">Morbus Iff</a>.
        <span class="backpedaling">These scripts are not affiliated with or created by the fine chaps and chapettes at Readmill.com.</span>
      </div>
    </div>
  </header>

  <div class="page-content container">
    <?php print $errors ? $errors : NULL; ?>
    <section class="layout-primary">
      <div class="layout-primary-column">
        <header>
          <h1>
            <span class="header-group-primary"><a href="<?php print $data->book->permalink_url; ?>"><?php print $data->book->title; ?></a></span>
            <span class="header-group-secondary"><?php print $data->book->author; ?></span>
          </h1>
        </header>
        <section id="highlights">
          <h1>Latest updates</h1>
          <?php
            foreach ($data->updates as $timestamp => $updates) {
              foreach ($updates as $update) {
                print '<article class="highlight">';
                print   '<blockquote>' . htmlentities($update['highlight']->content, ENT_COMPAT, "UTF-8") . '</blockquote>';
                print   '<section class="comments">'; // Be sure to enforce UTF-8 on all htmlentities() due to PHP 5.3 defaults.
                print     '<article class="comment">';
                print       '<a href="' . $update['user']->permalink_url . '" class="image">';
                print         '<img src="' . $update['user']->avatar_url . '" alt="' . $update['user']->fullname . '" />';
                print       '</a>';
                print       '<div class="content">';
                print         '<a href="' . $update['user']->permalink_url . '" class="fullname">' . $update['user']->fullname . '</a> ';
                print         '<p>' . htmlentities($update['content'], ENT_COMPAT, "UTF-8") . '</p>';
                print         '<aside class="metadata">';
                print           '<time class="timestamp" datetime="' . date(DATE_ISO8601, $timestamp) . '">' . date('D, d M Y h:i:s a', $timestamp) . '</time> &middot; ';
                print           '<a href="' . $update['permalink_url'] . '">' . $update['permalink_text'] . '</a>';
                print         '</aside>';
                print       '</div>';
                print     '</article>';
                print   '</section>';
                print '</article>';
              }
            }
          ?>
        </section>
      </div>
    </section>

    <aside class="layout-secondary">
      <div class="layout-secondary-column">
        <section class="book-cover">
          <a href="<?php print $data->book->permalink_url; ?>"><img class="book-cover" alt="Book cover for <?php print $data->book->title; ?>" src="<?php print str_replace('medium', 'original', $data->book->cover_url); ?>" /></a>
        </section>
        <section class="secondary-section">
          <form action="book-updates.php" accept-charset="UTF-8" method="get">
            <label for="form-title">Book title</label><input id="form-title" name="title" type="text" placeholder="<?php print htmlentities($match_title, ENT_COMPAT, "UTF-8"); ?>" required />
            <label for="form-author">Book author</label><input id="form-author" name="author" type="text" placeholder="<?php print htmlentities($match_author, ENT_COMPAT, "UTF-8"); ?>" required />
            <button>Load book</button>
          </form>
          <span class="warning"><strong>Be aware:</strong> If we've not seen this book before,
          or its data has expired, it might take 30 seconds or more before you'll get results.</span>
        </section>
        <section class="secondary-section">
          <h1>About this script</h1>
          <p><b>Book updates</b> shows the 50 most recent highlights and comments made to a book in HTML
          or <a href="<?php print htmlentities($url_for_rss, ENT_COMPAT, "UTF-8"); ?>">RSS</a>. For a
          user-centric approach, see <a href="user-updates.php">User updates</a>.</p>
        </section>
      </div>
    </aside>
  </div>

  <footer>
    <div class="container">
      <a href="http://www.disobey.com/d/code/readmilling/">Readmilling</a> is free software released
      under the <a href="http://www.gnu.org/licenses/quick-guide-gplv3.html">GPL v3</a> or later.
      <a href="https://github.com/morbus/readmilling">Browse the source code at Github.</a>
    </div>
  </footer>
</body>
</html>
